lib/logstash/outputs/webhdfs.rb

```diff
@@ -0,0 +1,278 @@
+  config :snappy_format, :validate => ["stream", "file"], :default => "stream"
+
+  # Remove @timestamp field. Hive does not like a leading "@", but we need @timestamp for path calculation.
+  config :remove_at_timestamp, :validate => :boolean, :default => true
+
+  public
+
+  def register
+    begin
+      require 'webhdfs'
+    rescue LoadError
+      @logger.error("Module webhdfs could not be loaded.")
+      raise
+    end
+    if @compression == "gzip"
+      begin
+        require "zlib"
+      rescue LoadError
+        @logger.error("Gzip compression selected but zlib module could not be loaded.")
+        raise
+      end
+    elsif @compression == "snappy"
+      begin
+        require "snappy"
+      rescue LoadError
+        @logger.error("Snappy compression selected but snappy module could not be loaded.")
+        raise
+      end
+    end
+    @files = {}
+    @client = prepare_client(@host, @port, @user)
+    # Test client connection.
+    begin
+      @client.list('/')
+    rescue => e
+      @logger.error("Webhdfs check request failed. (namenode: #{@client.host}:#{@client.port}, Exception: #{e.message})")
+      raise
+    end
+    buffer_initialize(
+      :max_items => @flush_size,
+      :max_interval => @idle_flush_time,
+      :logger => @logger
+    )
+  end # def register
+
+  public
+  def receive(event)
+    return unless output?(event)
+    buffer_receive(event)
+  end # def receive
+
+  def prepare_client(host, port, username)
+    client = WebHDFS::Client.new(host, port, username)
+    client.httpfs_mode = @use_httpfs
+    client.open_timeout = @open_timeout
+    client.read_timeout = @read_timeout
+    client.retry_known_errors = @retry_known_errors
+    client.retry_interval = @retry_interval if @retry_interval
+    client.retry_times = @retry_times if @retry_times
+    client
+  end
+
+  def flush(events=nil, teardown=false)
+    return if not events
+    # Avoid creating a new string for newline every time
+    newline = "\n".freeze
+    output_files = Hash.new { |hash, key| hash[key] = "" }
+    events.collect do |event|
+      path = event.sprintf(@path)
+      if @remove_at_timestamp
+        event.remove("@timestamp")
+      end
+      if @message_format
+        event_as_string = event.sprintf(@message_format)
+      else
+        event_as_string = event.to_json
+      end
+      event_as_string += newline unless event_as_string.end_with? newline
+      output_files[path] << event_as_string
+    end
+    output_files.each do |path, output|
+      if @compression == "gzip"
+        path += ".gz"
+        output = compress_gzip(output)
+      elsif @compression == "snappy"
+        path += ".snappy"
+        if @snappy_format == "file"
+          output = compress_snappy_file(output)
+        elsif
+          output = compress_snappy_stream(output)
+        end
+      end
+      write_tries = 0
+      while write_tries < @retry_times do
```

Comment:
You can implement all this into the write_data function by leveraging the retry notation of ruby, mode intel on http://www.tutorialspoint.com/ruby/ruby_loops.htm. This would avoid actually the break and next usage, plus simplifying the code.